common/utils/remoteServiceConnector.go

```diff
@@ -0,0 +1,109 @@
+package utils
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/auto"
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
+)
+
+// RemoteServiceConnector is a convenient struct to help implementing GetClientDataDeserializer().
+// Here is an example of usage when you want to serialize a Connection and a string.
+//
+//	 type ClientData struct {
+//	     MyString   string
+//	     Connection utils.Connection
+//	 }
+//
+//	 type MyRemoteService {
+//	     // This implement GetClientDataDeserializer() func(auto.UpResult) (*ClientData, error)
+//	     *RemoteServiceConnector[ClientData]
+//	 }
+//
+//		remoteServiceConnector := utils.NewRemoteServiceConnector(ctx, ClientData{})
+//		remoteServiceConnector.Register(stackKey, "Connection", connection)
+//		remoteServiceConnector.Register("test", "MyString", pulumi.String("Hello"))
+//
+//		return &MyRemoteService{
+//		    MyRemoteService: remoteServiceConnector
+//		}
+type RemoteServiceConnector[T any] struct {
+	ctx                  *pulumi.Context
+	stackKeyFieldNameMap map[string]string
+	value                T
+}
+
+// NewRemoteServiceConnector creates a new instance of RemoteServiceConnector[T].
+// initialValue is the initial value of the value returned by GetClientDataDeserializer.
+// initialValue can be used to return non pulumi.Input type.
+// T must be a struct.
+func NewRemoteServiceConnector[T any](ctx *pulumi.Context, initialValue T) *RemoteServiceConnector[T] {
+	return &RemoteServiceConnector[T]{
+		ctx:                  ctx,
+		value:                initialValue,
+		stackKeyFieldNameMap: make(map[string]string)}
+}
+
+// Register registers a field.
+// stackKeyName is the key in the stack name and should be unique
+// outputFieldName is the name of the field in the struct
+// value is the object to be saved in the stack
+func (s *RemoteServiceConnector[T]) Register(stackKeyName string, outputFieldName string, value pulumi.Input) {
```

Comment:
<!-- Suggestions propose improvements to the current subject. It’s important to be explicit and clear on what is being suggested and why it is an improvement. -->
**💬 suggestion**
Could we add an explicit "Check all fields of type T are registered" to be called to ensure all fields required by the deserialiser were registered ?

Reply:
The current implementation allows to pass an initial value for `T` in `NewRemoteServiceConnector`, so you can store non pulumi type like `int` for example.